File: veloce-backend/app/services/openai_analyze.py
from __future__ import annotations

import json
import logging
from typing import Literal

from openai import AsyncOpenAI
from pydantic import BaseModel, Field, ValidationError, field_validator, model_validator

from app.config import settings

logger = logging.getLogger(__name__)

ProjectCategory = Literal["Web App", "Mobile", "AI/ML", "Automation", "Integration"]


class ParsedAnalysis(BaseModel):
    features: list[str] = Field(min_length=1)
    category: ProjectCategory
    effort_hours_min: int = Field(gt=0)
    effort_hours_max: int = Field(gt=0)
    tech_stack: list[str] = Field(min_length=1)
    complexity: Literal[1, 2, 3, 4, 5]

    @field_validator("complexity", mode="before")
    @classmethod
    def complexity_int(cls, v: object) -> object:
        if isinstance(v, float):
            return int(v)
        return v

    @model_validator(mode="after")
    def effort_order(self) -> ParsedAnalysis:
        if self.effort_hours_max < self.effort_hours_min:
            raise ValueError("effort_hours_max must be >= effort_hours_min")
        return self


SYSTEM_PROMPT = """You analyze software project briefs for an agency. Respond with a single JSON object only (no markdown), using exactly these keys:
- "features": array of 3-8 short requirement strings derived from the brief
- "category": exactly one of: "Web App", "Mobile", "AI/ML", "Automation", "Integration"
- "effort_hours_min": positive integer (estimated engineering hours, lower bound)
- "effort_hours_max": positive integer (upper bound, must be >= effort_hours_min)
- "tech_stack": array of 3-8 suggested technologies
- "complexity": integer from 1 to 5 (5 = hardest)"""


async def analyze_brief(
    *,
    title: str,
    description_rich: str,
    budget_tier: str,
    timeline_urgency: str,
) -> tuple[ParsedAnalysis | None, str | None]:
    client = AsyncOpenAI(api_key=settings.openai_api_key)
    user_content = (
        f"Title: {title}\n"
        f"Budget tier: {budget_tier}\n"
        f"Timeline urgency: {timeline_urgency}\n\n"
        f"Description:\n{description_rich}"
    )
    last_err: str | None = None
    for attempt in range(3):
        try:
            resp = await client.chat.completions.create(
                model=settings.openai_model,
                messages=[
                    {"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user", "content": user_content},
                ],
                response_format={"type": "json_object"},
            )
            raw = resp.choices[0].message.content
            if not raw:
                last_err = "Empty model response"
                continue
            data = json.loads(raw)
            parsed = ParsedAnalysis.model_validate(data)
            return parsed, None
        except (json.JSONDecodeError, ValidationError, ValueError) as e:
            last_err = str(e)
            logger.warning("OpenAI parse attempt %s failed: %s", attempt + 1, e)
        except Exception as e:  # noqa: BLE001
            last_err = str(e)
            logger.warning("OpenAI API attempt %s failed: %s", attempt + 1, e)
    return None, last_err or "Analysis failed"

File: veloce-frontend/src/lib/server/openai-analyze.ts
import OpenAI, { APIError, RateLimitError } from "openai";
import { zodTextFormat } from "openai/helpers/zod";
import { z } from "zod";

import { getEnv } from "./env";
import { parsedAnalysisSchema } from "./schemas";

const INSTRUCTIONS = `You analyze software project briefs for an agency.
Return a structured analysis:
- features: 3-8 short requirement strings derived from the brief
- category: exactly one of: Web App, Mobile, AI/ML, Automation, Integration
- effort_hours_min / effort_hours_max: positive integers (engineering hours), max must be >= min
- tech_stack: 3-8 suggested technologies
- complexity: integer from 1 to 5 (5 = hardest)`;

const briefAnalysisFormat = zodTextFormat(parsedAnalysisSchema, "brief_analysis");

const MAX_ATTEMPTS = 3;
const BASE_BACKOFF_MS = 500;

export type ParsedAnalysis = z.infer<typeof parsedAnalysisSchema>;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function extractRefusal(response: {
  output?: Array<{
    type?: string;
    content?: Array<{ type?: string; refusal?: string }>;
  }>;
}): string | null {
  for (const item of response.output ?? []) {
    if (item.type !== "message" || !item.content) continue;
    for (const part of item.content) {
      if (part.type === "refusal" && typeof part.refusal === "string") {
        return part.refusal;
      }
    }
  }
  return null;
}

function backoffMsForAttempt(attempt: number, err: unknown): number {
  if (err instanceof RateLimitError) {
    return Math.min(8000, 2000 * (attempt + 1));
  }
  if (err instanceof APIError && err.status === 503) {
    return Math.min(6000, 1500 * (attempt + 1));
  }
  return BASE_BACKOFF_MS * (attempt + 1);
}

export async function analyzeBrief(input: {
  title: string;
  descriptionRich: string;
  budgetTier: string;
  timelineUrgency: string;
}): Promise<{ parsed: ParsedAnalysis | null; error: string | null }> {
  const env = getEnv();
  const client = new OpenAI({ apiKey: env.OPENAI_API_KEY });
  const userContent =
    `Title: ${input.title}\n` +
    `Budget tier: ${input.budgetTier}\n` +
    `Timeline urgency: ${input.timelineUrgency}\n\n` +
    `Description:\n${input.descriptionRich}`;

  let lastErr: string | null = null;
  let lastThrowable: unknown = null;

  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    try {
      const response = await client.responses.parse({
        model: env.OPENAI_MODEL,
        instructions: INSTRUCTIONS,
        input: [
          {
            role: "user",
            content: userContent,
          },
        ],
        text: {
          format: briefAnalysisFormat,
        },
      });

      lastThrowable = null;

      if (response.output_parsed) {
        return { parsed: response.output_parsed, error: null };
      }

      const refusal = extractRefusal(response);
      lastErr = refusal
        ? `Model refused: ${refusal}`
        : "Structured output was empty or could not be parsed";
      console.warn(
        `[analyzeBrief] attempt ${attempt + 1}/${MAX_ATTEMPTS}:`,
        lastErr,
      );
    } catch (e) {
      lastThrowable = e;
      lastErr = e instanceof Error ? e.message : String(e);
      if (e instanceof z.ZodError) {
        console.warn("[analyzeBrief] schema validation failed:", e.issues);
        lastErr = `Invalid structured output: ${e.issues.map((i) => i.message).join("; ")}`;
      } else if (e instanceof APIError) {
        console.warn(
          `[analyzeBrief] API error (status ${e.status ?? "?"}):`,
          e.message,
        );
      } else {
        console.warn("[analyzeBrief] request failed:", e);
      }
    }

    if (attempt < MAX_ATTEMPTS - 1) {
      await sleep(backoffMsForAttempt(attempt, lastThrowable));
    }
  }

  return { parsed: null, error: lastErr ?? "Analysis failed" };
}
